File: 0081_Search_in_Rotated_Sorted_Array_II/solution.java
package cn.huang.leetcode;

/*
81. Search in Rotated Sorted Array II
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e., [0,0,1,2,2,5,6] might become [2,5,6,0,0,1,2]).

You are given a target value to search. If found in the array return true, otherwise return false.

Example 1:

Input: nums = [2,5,6,0,0,1,2], target = 0
Output: true

Example 2:

Input: nums = [2,5,6,0,0,1,2], target = 3
Output: false

Follow up:

    This is a follow up problem to Search in Rotated Sorted Array, where nums may contain duplicates.
    Would this affect the run-time complexity? How and why?


 */
/*

相比 #33，有重复数字

 */
public class LC_0081_SearchinRotatedSortedArray_II {
    public boolean search(int[] nums, int target) {
        if (nums == null || nums.length == 0)
            return false;

        int start = 0, end = nums.length - 1;
        int mid;

        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (nums[mid] == target)
                return true;

            if (nums[mid] > nums[start])  // left side
            {
                if (nums[mid] >= target && nums[start] <= target)
                    end = mid;
                else
                    start = mid;
            } else if (nums[mid] < nums[start]) {
                if (nums[mid] <= target && nums[end] >= target)
                    start = mid;
                else
                    end = mid;
            } else  {
                // Important: Handle the worst case 111111110111;
                // nums[mid] == nums[start]
                start++;
            }
        }
        return (nums[start] == target || nums[end] == target);
    }
}

File: 0081_Search_in_Rotated_Sorted_Array_II/solution.cpp
/*
81. Search in Rotated Sorted Array II

https://leetcode.com/problems/search-in-rotated-sorted-array-ii/
[#33]
*/

#include <vector>

using namespace std;
/*
  相比 #33，有重复数字
*/
class Solution
{
public:
  bool search(vector<int> &nums, int target)
  {
    if (nums.size() == 0)
      return false;

    int l = 0;
    int r = nums.size() - 1;

    while (l <= r)
    {
      int mid = l + (r - l) / 2;
      if (nums[mid] == target)
      {
        return true;
      }

      if (nums[l] < nums[mid]) // mid located in left part, which is increasing
      {
        if (nums[l] <= target && target < nums[mid]) //  target located between left and mid
          r = mid - 1;
        else //  target is bigger than mid
          l = mid + 1;
      }
      else if (nums[l] > nums[mid]) // mid located in right part, which is increasing
      {
        if (nums[mid] < target && target <= nums[r]) //target located between mid and right
          l = mid + 1;
        else // target is less than mid
          r = mid - 1;
      }
      else
      {
        l++;
      }
    }
    return false;
  }
};